### DISCUSSION TITLE
Discussion:  Vulkan support

### DISCUSSION THREAD

**Discussion by ggarra13** [fltk/fltk#1021] [Ideas]
I missed Matthias Zoom meeting so I don't know if this was discussed.

**I am using FLTK**
 - Version 1.4.0
 - For version 1.x.x

**Describe the solution you'd like**
Vulkan, just as Wayland, is starting to gain support in the Linux VFX communities.  I played with vulkan in my 2D player, and there's no real advantage to it.  However, for 3D and compositing contents that may be different, as I saw in Houdini.  Bill Spitzak can probably vouch whether it is worth considering it or not.
In the VFX community, Houdini (SideFX 3d package) and Pixar's OpenUSD are starting to take advantage of it.  Another potential benefit of Vulkan is the posibility to start the move of FLTK to mobile devices, where GLFW shines.

**Describe alternatives you've considered**
Staying with OpenGL.

**Comment by MatthiasWM** [comment]
It's possible to implement a Vulkan window driver. I beleive that CMake in 1.4 made a huge stap, and it should now be possible to add more modules like `fltk-gl`. Not sure if any of the current maintainers have a desire to implement `Fl_Vulkan_Window` and maybe even a graphics driver, but I am sure if the community cooks one up that works, we can merge the code.

Vulkan seems to be the 3d library of choice for OpenXR to create virtual reality apps. FLTK in VR anyone?

**Reply by ggarra13** [reply]
> Vulkan seems to be the 3d library of choice for OpenXR to create virtual reality apps. FLTK in VR anyone?

OpenXR, in principle (have not tried it), supports all graphic APIs but Metal (albeit macOS now has a way to run vulkan, I understand).  For FLTK in VR or OpenXR we would need a way to create floating windows in 3D space.  FLTK 2D approach is ill-suited for that.  Game Engines look like a better proposition for VR/AR.
FLTK is better suited to go the Desktop / Web / Mobile approach.

**Reply by MatthiasWM** [reply]
> > Vulkan seems to be the 3d library of choice for OpenXR to create virtual reality apps. FLTK in VR anyone?
> 
> OpenXR, in principle (have not tried it), supports all graphic APIs but Metal (albeit macOS now has a way to run vulkan, I understand). For FLTK in VR or OpenXR we would need a way to create floating windows in 3D space. FLTK 2D approach is ill-suited for that. Game Engines look like a better proposition for VR/AR. FLTK is better suited to go the Desktop / Web / Mobile approach.

Yes, of course you are right. I did manage to create floating FLTK windows in VR, but it's just that, floating textured sheets. It fascinates me that 3d GUIs seem so logical in VR, but I still have to see a UX that is immersive *and* efficient at the same time. Plus I get motion sick easily 😂 .

**Comment by MatthiasWM** [comment]
PS: moving this to "Diskussions > Ideas"

**Comment by ggarra13** [comment]
I looked at the state of vulkan, as I saw blender is also planning to support it.  And so far I saw with the smallest KHR cubepp demo (on Linux):
         a) Window resizes make vulkan stop the redraws.  This is terrible.
         b) Vulkan provides simple wrappers around all window APIs, so porting FLTK to it is not as terrible as I thought.
         c) Swapchains are the main addition to the OpenGL pipeline and a cause of headaches.  I am still wrapping my head around them.
         d) Shaders must be compiled outside of the C++ code, with the glslang utility.  This sucks too.  Not sure how can you get something like OpenColorIO supported.
         e) Support on Windows and macOS depends on propietary binaries.

**Reply by ggarra13** [reply]
Ok.  I have a very ugly WIP support for Vulkan on my FLTK branch.  FLTK's drivers support made it easy.  Currently, it works on X11, Windows, macOS (with MoltenVK) and, in principle, Wayland (albeit currently a bug in FLTK? makes it crash).  I support both windows, embedded subwindows and resizing.
I have a demo based on GLFW's triangle and FLTK's GL shape demo which draws a triangle with a texture.  I am currently still learning the Vulkan API to understand how to best integrate it with FLTK.  There will probably be, at the start, three main classes, like Qt's approach to it (albeit simpler): Fl_Vk_Instance (an application level container called by the user at start), a Fl_Vk_Window_Driver (similar to Fl_Gl_Window_Driver hidden from the user) and Fl_Vk_Window (which is what the user will mostly interact with).  The Fl_Vk_Window_Driver calls each platforms' Fl_WinAPI_Vk_Window_Driver and similar to have it, at least for now, just create a surface.
From my look at the Vulkan API so far, it will not be possible to support the equivalent of gl_* functions, but it will likely be possible to have in the future other classes, like Fl_Vk_Mesh, Fl_Vk_Texture, etc.  This would be similar to the Khrono Group's framework, but tailored to FLTK.
On the Vulkan API's side you must rely too much on vkQueueWaitIdle() as FLTK's event loop is not multithreaded.
The one thing that I have that is worth looking so far is the CMake files.

Before I continue with this project, I have some questions:  am I stepping on someone's toes (ie. some FLTK developer is already working on it)?  If I become an FLTK developer, what's the developer's take on AI?  Should I avoid copying and pasting code from the AI to avoid license and copyright issues or noone cares anymore?

**Reply by rwtolbert** [reply]
@ggarra13 I'm very keen to learn more about your Vulkan work and help if I can. I've spent a couple of weeks building an app that needs 3D and it feels wrong to be using OpenGL in 2025. I know it "just works" but I can also see the day where macOS drops support for it.

**Reply by ggarra13** [reply]
@rwtolbert What platform are you working on and what are your hardware specs?  Also, let me know how you prefer to be contacted at ggarra13@gmail.com.

**Reply by erco77** [reply]
> ..Bill Spitzak can probably vouch..

Not sure if Bill is monitoring github "discussions" (his last commit to fltk was in 2005, before we switched to github), but he has replied on the groups fltkcoredev (last post Oct 2024) and fltkgeneral (Mar 2025), so if you want any vouching from him, might need to ask on one of those with an apropos Subject that will attract his attention.

**Reply by ggarra13** [reply]
@erco77 I sent two Vulkan questions with Bill in mind to fltkcoredev.

**Comment by rwtolbert** [comment]
I have access to M2 and M3 macOS, Windows Intel w/ an Nvidia card, Windows ARM and Fedora 42 Wayland with Intel or Nvidia.
